scapy/layers/ipsec.py

```diff
@@ -492,8 +492,8 @@ def decrypt(self, sa, esp, key, icv_size=None, esn_en=False, esn=0):
         nh = orb(data[-1])
 
         # then use padlen to determine data and padding
-        data = data[:len(data) - padlen - 2]
         padding = data[len(data) - padlen - 2: len(data) - 2]
+        data = data[:len(data) - padlen - 2]
```

Comment:
Seams reasonable to save the padding _before_ cutting it off, not afterwards. 😄

This sort of 'stupid' programming mistake can happen, and I doubt it is worth adding a dedicated test case for it. You'd end up testing every single assignment in your program.

Reply:
> I disagree. What's the point of fixing bugs if you don't test for regressions.

You are right. And I hope you also read my second post, where I qualified my first post.

What I meant to say: I'm not opposed to tests in general. But waiting for users to find bugs and then let the user write a regression test for this specific bug he found does not look like a scalable strategy to me for proactively reducing such types of errors imho. Also, not every reporter may have the skills to fiddle with the test suite, so it might end up in your responsibily to extend the tests. The reported bug clearly indicates that the current ESP test suite does not cover all members of all payloads, otherwise this error wouldn't have gone undetected.

One way to increase "payload coverage" would be to add more dissection tests of real packet captures, with full dissection and reassembly of the entire frame, comparing it to a handcrafted frame with all members explicitly specified. That's something I tried to do in the IKEv2 testsuite: I started with a pcap file, dissected it and compared the outcome manually with the Wireshark dissection. Then I converted it into a testcase by converting the packets into Python declarations using the `pkt.command()` output (and a lot of post-editing).

Comparing Scapy's dissections with other tools like Wireshark when writing the test would also find some types of bugs which you wouldn't find if you just do self-contained tests like crafting a packet manually in Scapy and then do some simple member tests in Scapy.


**Test run**

https://github.com/secdev/scapy/blob/e80b3d4457a30c361beb31e53e416cddb5bbbbb4/test/contrib/ikev2.uts#L1729-L1737

**Testcases**
https://github.com/secdev/scapy/blob/e80b3d4457a30c361beb31e53e416cddb5bbbbb4/test/contrib/ikev2.uts#L342-L416